File: apps/studio/src/bundle.ts
import { strToU8, zipSync } from "fflate";
import { auditTbx, canonicalJson, compareCodePoints } from "./tbxAudit";
import type { Engine, FieldPoint, FieldTable, GenerateRequest } from "./types";

function jsonBytes(value: unknown): Uint8Array {
  return strToU8(canonicalJson(value, true));
}

function jsonlBytes(rows: unknown[]): Uint8Array {
  return strToU8(rows.map((row) => canonicalJson(row)).join(""));
}

async function sha256(payload: Uint8Array | string): Promise<string> {
  const bytes = typeof payload === "string" ? strToU8(payload) : payload;
  const copy = new Uint8Array(bytes.byteLength);
  copy.set(bytes);
  const digest = await crypto.subtle.digest("SHA-256", copy.buffer);
  return [...new Uint8Array(digest)].map((value) => value.toString(16).padStart(2, "0")).join("");
}

async function contentHash(value: unknown): Promise<string> {
  return sha256(canonicalJson(value));
}

function statistics(points: FieldPoint[]) {
  const counts: Record<string, number> = {};
  for (const point of points) counts[point.classification] = (counts[point.classification] ?? 0) + 1;
  const mean = (field: "UI" | "NSS" | "S_e") => {
    const values = points.map((point) => point[field]).filter((value): value is number => value != null);
    return values.length ? Number((values.reduce((sum, value) => sum + value, 0) / values.length).toFixed(8)) : null;
  };
  return {
    point_count: points.length,
    classification_counts: Object.fromEntries(Object.entries(counts).sort(([left], [right]) => compareCodePoints(left, right))),
    mean_UI: mean("UI"),
    mean_NSS: mean("NSS"),
    mean_S_e: mean("S_e"),
    failure_count: 0,
  };
}

export async function exportBrowserBundle(table: FieldTable, request: GenerateRequest, engine: Engine) {
  const claimLevel = engine === "analytic" ? "ILLUSTRATIVE_ANALYTIC" : "COMPUTED_DYNAMICAL";
  const kernelId = `browser.${engine}.v1`;
  const nullCount = engine === "analytic" ? 0 : 12;
  const specification = {
    schema_version: "1.0.0",
    engine,
    seed: request.seed,
    domain_id: engine === "analytic" ? "analytic-browser-preview" : "synthetic-ring-14",
    claim_level: claimLevel,
    grid: { width: table.width, height: table.height },
    parameters: {
      power: request.power,
      max_iterations: request.maxIterations,
      recovery_steps: request.recoverySteps,
    },
    classification_rules: {
      separation_threshold: 0.08,
      nss_threshold: 1,
      survival_threshold: 0.6,
      recovery_threshold: 0.78,
      escape_threshold: 0.46,
    },
    null_policy: engine === "analytic"
      ? { kind: "none", count: 0, seed: request.seed }
      : { kind: "preserve_multiset_shuffle", count: nullCount, seed: request.seed },
  };
  const specificationSha256 = await contentHash(specification);
  const runId = `run-${(await contentHash({ specification, kernel_id: kernelId, domain_sha256: null })).slice(0, 16)}`;
  const tableForBundle = {
    schema_version: "1.0.0",
    width: table.width,
    height: table.height,
    points: table.points,
  };
  const summary = statistics(table.points);
  const winnerCounts: Record<string, number> = {};
  for (const point of table.points) {
    if (point.winner_N != null) winnerCounts[String(point.winner_N)] = (winnerCounts[String(point.winner_N)] ?? 0) + 1;
  }
  const nullRegistry = Array.from({ length: nullCount }, (_, index) => ({
    null_id: `browser-null-${String(index).padStart(2, "0")}`,
    policy: "preserve_multiset_shuffle",
    seed: request.seed + index,
  }));
  const files: Record<string, Uint8Array> = {
    "run_spec.json": jsonBytes(specification),
    "ontology.json": jsonBytes({
      schema_version: "1.0.0",
      terms: { omega: "ordered ladder state", T_e: "first observed/null separation", S_e: "persistence after emergence", winner_N: "closure-mode label" },
      non_equivalences: ["closure != emergence", "survival != closure", "interpolation != observation"],
    }),
    "claim_boundary.json": jsonBytes({
      claim_level: claimLevel,
      permitted_interpretations: ["browser-side reproducible preview"],
      excluded_interpretations: ["external validation", "causal authority", "scientific use without CPU regeneration"],
      experimental_tags: ["browser_preview"],
      independent_verifier_status: "not_supplied",
    }),
    "visual_encoding.json": jsonBytes({
      position: { x: engine === "analytic" ? "complex_real" : "order_mutation_strength", y: engine === "analytic" ? "complex_imaginary" : "anchoring_alpha", z: "S_e" },
      color: "selected_by_user",
      interpolation: { method: "bilinear", used_for_metrics: false },
    }),
    "scene_recipe.json": jsonBytes({ viewer: "TORUS Field Studio", raw_samples_visible: true }),
    "provenance/sources.jsonl": jsonlBytes([{ source_id: specification.domain_id, kind: "browser_preview", engine }]),
    "provenance/transformations.jsonl": jsonlBytes([{
      transformation_id: kernelId,
      software_version: "0.2.1",
      seed: request.seed,
      specification_sha256: specificationSha256,
    }]),
    "provenance/verification_receipts.jsonl": new Uint8Array(),
    "registry/parent_registry.json": jsonBytes({
      parent_id: table.points[0]?.parent_id ?? "none",
      kind: engine === "analytic" ? "declared_initial_state" : "browser_synthetic_fixture",
      claim_authority: claimLevel,
    }),
    "registry/null_registry.json": jsonBytes(nullRegistry),
    "tables/field_points.json": strToU8(canonicalJson(tableForBundle)),
    "tables/metrics_by_N.json": jsonBytes({ winner_N_counts: winnerCounts, summary }),
    "audit/audit.json": jsonBytes({
      status: "browser_preview_unverified",
      kernel_authority: "deterministic_browser_preview",
      failure_preservation: "complete",
      classification_precedes_rendering: true,
    }),
    "audit/failure_ledger.jsonl": new Uint8Array(),
  };
  const entries = await Promise.all(Object.entries(files).map(async ([path, payload]) => ({ path, sha256: await sha256(payload), bytes: payload.byteLength })));
  files["audit/SHA256SUMS.txt"] = strToU8(entries
    .sort((left, right) => compareCodePoints(left.path, right.path))
    .map((entry) => `${entry.sha256}  ${entry.path}\n`).join(""));
  entries.push({
    path: "audit/SHA256SUMS.txt",
    sha256: await sha256(files["audit/SHA256SUMS.txt"]),
    bytes: files["audit/SHA256SUMS.txt"].byteLength,
  });
  files["manifest.json"] = jsonBytes({
    tbx_version: "1.0.0",
    run_id: runId,
    claim_level: claimLevel,
    kernel_id: kernelId,
    specification_sha256: specificationSha256,
    statistics: summary,
    files: entries.sort((left, right) => compareCodePoints(left.path, right.path)),
  });
  const blob = new Blob([zipSync(files, { level: 9 }) as BlobPart], { type: "application/zip" });
  const link = document.createElement("a");
  link.href = URL.createObjectURL(blob);
  link.download = `${runId}.tbx.zip`;
  link.click();
  URL.revokeObjectURL(link.href);
}

export async function importBundle(file: File): Promise<FieldTable> {
  const bytes = new Uint8Array(await file.arrayBuffer());
  const audit = await auditTbx(bytes);
  if (!audit.valid || !audit.table || !audit.manifest || !audit.specification) {
    const summary = audit.errors.slice(0, 3).join("; ");
    throw new Error(`TBX audit rejected [${audit.issueCodes.join(", ")}]: ${summary}`);
  }
  const engine = audit.specification.engine;
  return {
    ...audit.table,
    source: "tbx_import",
    runId: audit.manifest.run_id,
    claimLevel: audit.manifest.claim_level,
    engine: engine === "analytic" || engine === "local_brot" || engine === "tld" ? engine : undefined,
    auditCheckedFiles: audit.checkedFiles,
    tld: audit.tld,
    heldout: audit.heldout,
  };
}

File: python/torusbrot/bundle.py
from __future__ import annotations

import csv
import hashlib
import json
import zipfile
from collections import Counter
from dataclasses import dataclass
from pathlib import Path
from typing import Any

from .audit import audit_bundle, read_bundle_members
from .models import (
    AuditReport,
    ClaimLevel,
    DomainPack,
    FailureRecord,
    FieldPoint,
    RunSpec,
    canonical_json,
    content_hash,
)


def _jsonl(rows: list[dict[str, Any]]) -> bytes:
    return b"".join(canonical_json(row) for row in rows)


def _sha256(payload: bytes) -> str:
    return hashlib.sha256(payload).hexdigest()


@dataclass
class FieldResult:
    specification: RunSpec
    points: list[FieldPoint]
    kernel_id: str
    claim_level: ClaimLevel
    run_id: str
    domain: DomainPack | None
    null_registry: list[dict[str, Any]]
    failures: list[FailureRecord]

    @classmethod
    def from_run(
        cls,
        *,
        specification: RunSpec,
        points: list[FieldPoint],
        kernel_id: str,
        claim_level: ClaimLevel,
        domain: DomainPack | None,
        null_registry: list[dict[str, Any]],
        failures: list[FailureRecord] | None = None,
    ) -> FieldResult:
        identity = {
            "specification": specification.to_dict(),
            "kernel_id": kernel_id,
            "domain_sha256": domain.sha256 if domain else None,
        }
        return cls(
            specification=specification,
            points=points,
            kernel_id=kernel_id,
            claim_level=claim_level,
            run_id=f"run-{content_hash(identity)[:16]}",
            domain=domain,
            null_registry=null_registry,
            failures=failures or [],
        )

    @property
    def statistics(self) -> dict[str, Any]:
        classifications = Counter(point.classification for point in self.points)
        count = max(len(self.points), 1)
        return {
            "point_count": len(self.points),
            "classification_counts": dict(sorted(classifications.items())),
            "mean_UI": round(sum(point.UI for point in self.points) / count, 8),
            "mean_NSS": round(sum(point.NSS for point in self.points) / count, 8),
            "mean_S_e": round(sum(point.S_e for point in self.points) / count, 8),
            "failure_count": len(self.failures),
        }

    def _members(self) -> dict[str, bytes]:
        spec = self.specification.to_dict()
        analytic = self.specification.engine == "analytic"
        points = [point.to_dict() for point in self.points]
        winner_counts = Counter(str(point.winner_N) for point in self.points if point.winner_N)
        parent_registry = (
            {
                "parent_id": self.domain.domain_id,
                "title": self.domain.title,
                "ladder": list(self.domain.ladder),
                "domain_sha256": self.domain.sha256,
                "claim_authority": self.domain.claim_authority.name,
            }
            if self.domain
            else {"parent_id": "analytic-origin-z0", "kind": "declared_initial_state"}
        )
        permitted = (
            ["visual analogy", "orbit comparison", "reproducible analytic exploration"]
            if analytic
            else ["registered computation", "parent-versus-matched-null comparison"]
        )
        excluded = [
            "external validation without an independent verification receipt",
            "causal inference from visual appearance",
            "count, repair, release, or terminal authority",
        ]
        members: dict[str, bytes] = {
            "run_spec.json": canonical_json(spec, pretty=True),
            "ontology.json": canonical_json(
                {
                    "schema_version": "1.0.0",
                    "terms": {
                        "omega": "ordered ladder state",
                        "T_e": "first registered depth of observed/null separation",
                        "S_e": "persistence of separation after emergence",
                        "winner_N": "harmonic closure-mode label",
                    },
                    "non_equivalences": [
                        "closure != emergence",
                        "survival != closure",
                        "escape != failure",
                        "interpolation != observation",
                    ],
                },
                pretty=True,
            ),
            "claim_boundary.json": canonical_json(
                {
                    "claim_level": self.claim_level.name,
                    "permitted_interpretations": permitted,
                    "excluded_interpretations": excluded,
                    "experimental_tags": [],
                    "independent_verifier_status": "not_supplied",
                },
                pretty=True,
            ),
            "visual_encoding.json": canonical_json(
                {
                    "position": {
                        "x": "complex_real" if analytic else "order_mutation_strength",
                        "y": "complex_imaginary" if analytic else "anchoring_alpha",
                        "z": "S_e",
                    },
                    "color": "iterations" if analytic else "NSS",
                    "opacity": "UI",
                    "contour": None if analytic else "winner_N",
                    "interpolation": {"method": "bilinear", "used_for_metrics": False},
                    "scale_policy": "frozen_domain_standard",
                },
                pretty=True,
            ),
            "scene_recipe.json": canonical_json(
                {
                    "viewer": "TORUS Field Studio",
                    "default_view": "field",
                    "raw_samples_visible": True,
                    "missing_regions_visible": True,
                    "palette": "spectral-claim-safe-v1",
                },
                pretty=True,
            ),
            "provenance/sources.jsonl": _jsonl(
                [
                    {
                        "source_id": self.domain.domain_id if self.domain else "analytic-equation",
                        "sha256": self.domain.sha256 if self.domain else content_hash(spec),
                        "kind": self.domain.source.get("kind") if self.domain else "declared_model",
                        "details": (
                            self.domain.source
                            if self.domain
                            else {"equation": "z[n+1]=z[n]^p+c"}
                        ),
                    }
                ]
            ),
            "provenance/transformations.jsonl": _jsonl(
                [
                    {
                        "transformation_id": self.kernel_id,
                        "software_version": "0.2.0",
                        "seed": self.specification.seed,
                        "specification_sha256": self.specification.sha256,
                    }
                ]
            ),
            "provenance/verification_receipts.jsonl": b"",
            "registry/parent_registry.json": canonical_json(parent_registry, pretty=True),
            "registry/null_registry.json": canonical_json(self.null_registry, pretty=True),
            "tables/field_points.json": canonical_json(
                {
                    "schema_version": "1.0.0",
                    "width": self.specification.grid.width,
                    "height": self.specification.grid.height,
                    "points": points,
                }
            ),
            "tables/metrics_by_N.json": canonical_json(
                {
                    "winner_N_counts": dict(sorted(winner_counts.items())),
                    "summary": self.statistics,
                },
                pretty=True,
            ),
            "audit/audit.json": canonical_json(
                {
                    "status": "self_checked",
                    "kernel_authority": "deterministic_cpu_reference",
                    "failure_preservation": "complete",
                    "classification_precedes_rendering": True,
                },
                pretty=True,
            ),
            "audit/failure_ledger.jsonl": _jsonl(
                [failure.to_dict() for failure in self.failures]
            ),
        }
        sums = "".join(
            f"{_sha256(payload)}  {path}\n" for path, payload in sorted(members.items())
        )
        members["audit/SHA256SUMS.txt"] = sums.encode()
        return members

    def export_tbx(self, destination: str | Path) -> Path:
        target = Path(destination)
        members = self._members()
        manifest = {
            "tbx_version": "1.0.0",
            "run_id": self.run_id,
            "claim_level": self.claim_level.name,
            "kernel_id": self.kernel_id,
            "specification_sha256": self.specification.sha256,
            "statistics": self.statistics,
            "files": [
                {"path": path, "sha256": _sha256(payload), "bytes": len(payload)}
                for path, payload in sorted(members.items())
            ],
        }
        manifest_payload = canonical_json(manifest, pretty=True)
        if target.name.endswith(".zip"):
            target.parent.mkdir(parents=True, exist_ok=True)
            with zipfile.ZipFile(
                target, "w", compression=zipfile.ZIP_DEFLATED, compresslevel=9
            ) as archive:
                ordered_members = [("manifest.json", manifest_payload), *sorted(members.items())]
                for path, payload in ordered_members:
                    info = zipfile.ZipInfo(path, date_time=(1980, 1, 1, 0, 0, 0))
                    info.create_system = 3
                    info.compress_type = zipfile.ZIP_DEFLATED
                    info.external_attr = 0o100644 << 16
                    archive.writestr(info, payload)
        else:
            target.mkdir(parents=True, exist_ok=True)
            (target / "manifest.json").write_bytes(manifest_payload)
            for path, payload in members.items():
                member = target / path
                member.parent.mkdir(parents=True, exist_ok=True)
                member.write_bytes(payload)
        return target

    def audit(self, source: str | Path | None = None) -> AuditReport:
        if source is None:
            import tempfile

            with tempfile.TemporaryDirectory(prefix="torusbrot-audit-") as temporary:
                target = self.export_tbx(Path(temporary) / "run.tbx")
                return audit_bundle(target)
        return audit_bundle(source)


def read_field_table(source: str | Path) -> dict[str, Any]:
    report = audit_bundle(source)
    if not report.valid:
        raise ValueError("Bundle audit failed: " + "; ".join(report.errors))
    members = read_bundle_members(source)
    try:
        return json.loads(members["tables/field_points.json"])
    except KeyError as error:
        raise ValueError("Bundle does not contain tables/field_points.json") from error


def export_field_csv(source: str | Path, destination: str | Path) -> Path:
    table = read_field_table(source)
    target = Path(destination)
    target.parent.mkdir(parents=True, exist_ok=True)
    points = table["points"]
    excluded = {"trace"}
    headers = [key for key in points[0] if key not in excluded] if points else []
    with target.open("w", newline="", encoding="utf-8") as handle:
        writer = csv.DictWriter(handle, fieldnames=headers)
        writer.writeheader()
        for point in points:
            writer.writerow({key: point[key] for key in headers})
    return target


def copy_field_json(source: str | Path, destination: str | Path) -> Path:
    target = Path(destination)
    target.parent.mkdir(parents=True, exist_ok=True)
    target.write_bytes(canonical_json(read_field_table(source), pretty=True))
    return target


def compare_bundles(left: str | Path, right: str | Path) -> dict[str, Any]:
    left_members = read_bundle_members(left)
    right_members = read_bundle_members(right)
    left_manifest = json.loads(left_members["manifest.json"])
    right_manifest = json.loads(right_members["manifest.json"])
    left_points, right_points = read_field_table(left)["points"], read_field_table(right)["points"]
    left_classes = Counter(point["classification"] for point in left_points)
    right_classes = Counter(point["classification"] for point in right_points)
    return {
        "left_run_id": left_manifest["run_id"],
        "right_run_id": right_manifest["run_id"],
        "identical_specification": left_manifest.get("specification_sha256")
        == right_manifest.get("specification_sha256"),
        "point_count_delta": len(right_points) - len(left_points),
        "classification_delta": {
            key: right_classes[key] - left_classes[key]
            for key in sorted(left_classes.keys() | right_classes.keys())
        },
    }
